src/CloudNative.CloudEvents/Http/HttpUtilities.cs

```diff
@@ -69,42 +70,242 @@ public static class HttpUtilities
         {
-            // Apply https://tools.ietf.org/html/rfc3986#section-2.4
-            // then https://tools.ietf.org/html/rfc7230#section-3.2.6
-            // FIXME: Do this properly, including handling values that contain commas.
-            // It's possible that the underlying HTTP libraries do this already, of course...
-            // need to use Wireshark to check exactly what's happening on the wire.
-            // (Test with headers containing commas, double quotes, surrogate pairs etc)
-
-            // For the moment, encode:
-            // % as %25
-            // all non-ASCII characters, not worrying about surrogate pairs...
-            var builder = new StringBuilder();
-            foreach (char c in value)
+            Validation.CheckNotNull(value, nameof(value));
+            const string HexDigits = "0123456789ABCDEF";
+
+            // TODO: Use stackalloc, or perform the UTF-8 encoding manually.
+            // (It's pretty simple to do.)
+
+            // Buffer to use for UTF-8 encoding.
+            byte[] buffer = null;
+
+            var builder = new StringBuilder(value.Length);
+            for (int i = 0; i < value.Length; i++)
             {
-                if (c < ' ' || c > '\u007f' || c == '%')
+                int codePoint = char.ConvertToUtf32(value, i);
+
+                // Encode: non-printable-ASCII, non-ASCII, space, percent, double-quote.
+                // Start by handling the ASCII special cases which are all hard-coded (and most common)
+                switch (codePoint)
                 {
-                    builder.Append(WebUtility.UrlEncode(c.ToString()));
+                    case ' ':
+                        builder.Append("%20");
+                        continue;
+                    case '"':
+                        builder.Append("%22");
+                        continue;
+                    case '%':
+                        builder.Append("%25");
+                        continue;
                 }
-                else
+                // Now check for simple ASCII that doesn't need to be encoded
+                if (codePoint > 0x20 && codePoint < 0x7f)
                 {
-                    builder.Append(c);
+                    builder.Append((char) codePoint);
+                    continue;
+                }
+
+                // Full-on UTF-8 encoding now...
+                if (buffer is null)
+                {
+                    buffer = new byte[4];
+                }
+                int byteCount = Encoding.UTF8.GetBytes(value, i, codePoint > 0xffff ? 2 : 1, buffer, 0);
+                for (int j = 0; j < byteCount; j++)
+                {
+                    builder.Append('%');
+                    builder.Append(HexDigits[buffer[j] >> 4]);
+                    builder.Append(HexDigits[buffer[j] & 0x0f]);
+                }
+
+                // Ensure we iterate over code points, not UTF-16 code units.
+                if (codePoint > 0xffff)
```

Comment:
Would it be worth consolidating this check with the one on line 121.

Reply:
Done.